external/kafka/src/main/scala/org/apache/spark/rdd/kafka/KafkaRDD.scala

```diff
@@ -0,0 +1,199 @@
+        // to minimize number of kafka metadata requests
+        private def connectLeader: SimpleConsumer = {
+          if (context.attemptNumber > 0) {
+            kc.connectLeader(part.topic, part.partition).fold(
+              errs => throw new Exception(
+                s"Couldn't connect to leader for topic ${part.topic} ${part.partition}: " +
+                  errs.mkString("\n")),
+              consumer => consumer
+            )
+          } else {
+            kc.connect(part.host, part.port)
+          }
+        }
+
+        private def handleErr(resp: FetchResponse) {
+          if (resp.hasError) {
+            val err = resp.errorCode(part.topic, part.partition)
+            if (err == ErrorMapping.LeaderNotAvailableCode ||
+              err == ErrorMapping.NotLeaderForPartitionCode) {
+              log.error(s"Lost leader for topic ${part.topic} partition ${part.partition}, " +
+                s" sleeping for ${kc.config.refreshLeaderBackoffMs}ms")
+              Thread.sleep(kc.config.refreshLeaderBackoffMs)
+            }
+            // Let normal rdd retry sort out reconnect attempts
+            throw ErrorMapping.exceptionFor(err)
+          }
+        }
+
+        override def close() = consumer.close()
+
+        override def getNext: R = {
+          if (iter == null || !iter.hasNext) {
+            val req = new FetchRequestBuilder().
+              addFetch(part.topic, part.partition, requestOffset, kc.config.fetchMessageMaxBytes).
+              build()
+            val resp = consumer.fetch(req)
+            handleErr(resp)
+            // kafka may return a batch that starts before the requested offset
+            iter = resp.messageSet(part.topic, part.partition)
+              .iterator
+              .dropWhile(_.offset < requestOffset)
+          }
+          if (!iter.hasNext) {
+            assert(requestOffset == part.untilOffset,
+              s"ran out of messages before reaching ending offset ${part.untilOffset} " +
+                s"for topic ${part.topic} partition ${part.partition} start ${part.fromOffset}." +
+                " This should not happen, and indicates that messages may have been lost")
+            finished = true
+            null.asInstanceOf[R]
+          } else {
+            val item = iter.next
+            if (item.offset >= part.untilOffset) {
+              assert(item.offset == part.untilOffset,
+                s"got ${item.offset} > ending offset ${part.untilOffset} " +
+                  s"for topic ${part.topic} partition ${part.partition} start ${part.fromOffset}." +
+                  " This should not happen, and indicates a message may have been skipped")
+              finished = true
+              null.asInstanceOf[R]
+            } else {
+              requestOffset = item.nextOffset
+              messageHandler(new MessageAndMetadata(
+                part.topic, part.partition, item.message, item.offset, keyDecoder, valueDecoder))
+            }
```

Comment:
I agree that's one of the ugliest parts of the code, but unfortunately I
think those are the cases that need to be handled.  I'll see if there's a
way to flatten it.

On Mon, Jan 26, 2015 at 5:06 PM, Reynold Xin <notifications@github.com>
wrote:

> In external/kafka/src/main/scala/org/apache/spark/rdd/kafka/KafkaRDD.scala
> <https://github.com/apache/spark/pull/3798#discussion_r23573027>:
>
> > +            finished = true
> > +            null.asInstanceOf[R]
> > +          } else {
> > +            val item = iter.next
> > +            if (item.offset >= part.untilOffset) {
> > +              assert(item.offset == part.untilOffset,
> > +                s"got ${item.offset} > ending offset ${part.untilOffset} " +
> > +                  s"for topic ${part.topic} partition ${part.partition} start ${part.fromOffset}." +
> > +                  " This should not happen, and indicates a message may have been skipped")
> > +              finished = true
> > +              null.asInstanceOf[R]
> > +            } else {
> > +              requestOffset = item.nextOffset
> > +              messageHandler(new MessageAndMetadata(
> > +                part.topic, part.partition, item.message, item.offset, keyDecoder, valueDecoder))
> > +            }
>
> this level of conditional / structure nesting seems scary. Any chance we
> can reduce this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3798/files#r23573027>.
>